tests/gtests/internals/test_sdpa.cpp

```diff
@@ -0,0 +1,1916 @@
+// with -inf.
+void fill_mask(std::vector<float> &mask, memory::desc desc) {
+    size_t seq_len = desc.get_dims()[3];
+    size_t query_num = desc.get_dims()[2];
+    size_t batches = desc.get_dims()[1] * desc.get_dims()[0];
+    for (size_t b = 0; b < batches; b++) {
+        for (size_t q = 0; q < query_num; q++) {
+            for (size_t i = 0; i < seq_len; i++) {
+                if (i <= q) {
+                    mask[b * query_num * seq_len + q * seq_len + i] = 0;
+                    // = (float)i + (float)q / 100.f;
+                } else {
+                    mask[b * query_num * seq_len + q * seq_len + i]
+                            = -1 * std::numeric_limits<float>::infinity();
+                    //= -((float)i + (float)q / 100.f);
+                }
+            }
+        }
+    }
+}
+
+void fill_causal_mask(std::vector<float> &mask, memory::desc desc) {
+    size_t seq_len = desc.get_dims()[3];
+    size_t query_num = desc.get_dims()[2];
+    size_t batches = desc.get_dims()[1] * desc.get_dims()[0];
+    for (size_t b = 0; b < batches; b++) {
+        for (size_t q = 0; q < query_num; q++) {
+            for (size_t i = 0; i < seq_len; i++) {
+                if (i <= q) {
+                    mask[b * query_num * seq_len + q * seq_len + i] = 0;
+                    // = (float)i + (float)q / 100.f;
+                } else {
+                    mask[b * query_num * seq_len + q * seq_len + i]
+                            = -1 * std::numeric_limits<float>::infinity();
+                    //= -((float)i + (float)q / 100.f);
+                }
+            }
+        }
+    }
+}
+
+void print_mem(const dnnl::memory &mem, std::string name = "") {
+    auto eng = mem.get_engine();
+    dnnl::stream s(eng);
+    s.wait();
+    auto desc = mem.get_desc();
+    auto dims = desc.get_dims();
+    auto strides = desc.get_strides();
+    printf("%sbegin\n", name.c_str());
+    printf("dims   : %6ld %6ld %6ld %6ld\n", dims[0], dims[1], dims[2],
+            dims[3]);
+    printf("strides: %6ld %6ld %6ld %6ld\n", strides[0], strides[1], strides[2],
+            strides[3]);
+    if (mem.get_desc().get_data_type() == dnnl_bf16) { printf("bf16\n"); }
+    void *mapped_ptr_ = (void *)mem.map_data();
+    printf("        i:    ");
+    for (int i = 0; i < dims[3]; i++) {
+        switch ((int)desc.get_data_type()) {
+            case dnnl_u4:
+            case dnnl_s4: printf("%4d", i); break;
+            case dnnl_u8:
+            case dnnl_s8: printf("%4d", i); break;
+            case dnnl_f32:
+            case dnnl_bf16:
+            case dnnl_f16: printf("%9d", i); break;
+        }
+    }
+    printf("\n");
+
+    switch ((int)desc.get_data_type()) {
+        case dnnl_u4:
+        case dnnl_s4: {
+            char *mapped_ptr = (char *)mapped_ptr_;
+            for (int l = 0; l < dims[0]; l++) {
+                for (int k = 0; k < dims[1]; k++) {
+                    for (int j = 0; j < dims[2]; j++) {
```

Comment:
Minor: this can be re-written in a more compact form of:
```
for_()
for_()
for () {
...
}
```
It's a custom syntax to improve formatting.